src/main/java/com/patchworkmc/Patchwork.java

```diff
@@ -53,79 +55,80 @@
-		TsrgMappings mappings = new TsrgMappings(classes, intermediary);
-
-		if (!voldemapTiny.exists()) {
-			TinyWriter tinyWriter = new TinyWriter("official", "srg");
-			mappings.load(tinyWriter);
-			String tiny = tinyWriter.toString();
-			Files.write(voldemapTiny.toPath(), tiny.getBytes(StandardCharsets.UTF_8));
+		try {
+			patchworkGreyscaleIcon = Files.readAllBytes(new File(Patchwork.class.getResource("/patchwork-icon-greyscale.png").getPath()).toPath());
+		} catch (IOException ex) {
+			LOGGER.thrown(LogLevel.FATAL, ex);
 		}
+	}
 
-		File voldemapBridged = new File(current, "data/mappings/voldemap-bridged-" + version + ".tiny");
-		IMappingProvider bridged;
-
-		if (!voldemapBridged.exists()) {
-			LOGGER.trace("Generating bridged (srg -> intermediary) tiny mappings");
-
-			TinyWriter tinyWriter = new TinyWriter("srg", "intermediary");
-			bridged = new BridgedMappings(mappings, intermediary);
-			bridged.load(tinyWriter);
-
-			Files.write(voldemapBridged.toPath(), tinyWriter.toString().getBytes(StandardCharsets.UTF_8));
-		} else {
-			LOGGER.trace("Using cached bridged (srg -> intermediary) tiny mappings");
+	public Patchwork(Path inputDir, Path outputDir, Path dataDir, Path tempDir, IMappingProvider... mappings) {
+		this.inputDir = inputDir;
+		this.outputDir = outputDir;
+		this.dataDir = dataDir;
+		this.tempDir = tempDir;
+		this.mappingProviders = Arrays.asList(mappings);
 
-			bridged = TinyUtils.createTinyMappingProvider(voldemapBridged.toPath(), "srg", "intermediary");
+		if (this.mappingProviders.isEmpty()) {
+			throw new IllegalArgumentException("Must have at least one IMappingProvider!");
 		}
 
-		NaiveRemapper naiveRemapper = new NaiveRemapper(bridged);
-		Files.createDirectories(currentPath.resolve("input"));
-		Files.createDirectories(currentPath.resolve("output"));
-		Stream<Path> inputWalk = Files.walk(currentPath.resolve("input"));
-		inputWalk.forEach(file -> {
-			if (!file.toString().endsWith(".jar")) {
-				return;
-			}
-
-			String modName = file.getFileName().toString().replaceAll(".jar", "");
-
-			LOGGER.info("=== Transforming " + modName + " ===");
+		IMappingProvider mainMappings = mappingProviders.get(0);
+		this.naiveRemapper = new NaiveRemapper(mainMappings);
+		this.manifestRemapper = new ManifestRemapper(mainMappings);
+	}
 
-			try {
-				transformMod(currentPath, file, currentPath.resolve("output"), modName, bridged, naiveRemapper);
-			} catch (Exception e) {
-				LOGGER.error("Transformation failed, going on to next mod: ");
+	public int patchAndFinish() throws IOException {
+		if (this.closed) {
+			throw new IllegalStateException("Cannot being patching: Already patched all mods!");
+		}
+		AtomicInteger count = new AtomicInteger();
+
+		try (Stream<Path> inputFiles = Files.walk(inputDir).filter(file -> file.toString().endsWith(".jar"))) {
+			inputFiles.peek(jarPath -> {
```

Comment:
I would prefer for this to just become a normal `for` loop, it looks like the logic in peek could just be in forEach to begin with

Reply:
forEach can't be chained on streams, which is why it's a peek call. I'll convert them to normal loops instead.